libs/@local/hql/diagnostics/src/encoding.rs

```diff
@@ -0,0 +1,200 @@
+use core::{
+    fmt::{self, Display},
+    str::FromStr,
+};
+
+use anstyle::{Ansi256Color, AnsiColor};
+use serde::{
+    de::{value::StrDeserializer, Visitor},
+    ser::SerializeSeq,
+    Deserialize, Serialize,
+};
+use serde_with::{DeserializeAs, SerializeAs};
+
+struct AnsiParseError;
+
+impl Display for AnsiParseError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("invalid ANSI color, expected:")?;
+
+        for (index, color) in Ansi::ALL.into_iter().enumerate() {
+            if index != 0 {
+                f.write_str(", or")?;
+            }
+
+            write!(f, " `{color}`")?;
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(
+    Debug,
+    Copy,
+    Clone,
+    PartialEq,
+    Eq,
+    Hash,
+    serde_with::SerializeDisplay,
+    serde_with::DeserializeFromStr,
+)]
+struct Ansi(AnsiColor);
+
+impl Ansi {
+    const ALL: [Self; 16] = [
+        Self(AnsiColor::Black),
+        Self(AnsiColor::Red),
+        Self(AnsiColor::Green),
+        Self(AnsiColor::Yellow),
+        Self(AnsiColor::Blue),
+        Self(AnsiColor::Magenta),
+        Self(AnsiColor::Cyan),
+        Self(AnsiColor::White),
+        Self(AnsiColor::BrightBlack),
+        Self(AnsiColor::BrightRed),
+        Self(AnsiColor::BrightGreen),
+        Self(AnsiColor::BrightYellow),
+        Self(AnsiColor::BrightBlue),
+        Self(AnsiColor::BrightMagenta),
+        Self(AnsiColor::BrightCyan),
+        Self(AnsiColor::BrightWhite),
+    ];
+}
+
+impl Display for Ansi {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self.0 {
+            AnsiColor::Black => f.write_str("black"),
+            AnsiColor::Red => f.write_str("red"),
+            AnsiColor::Green => f.write_str("green"),
+            AnsiColor::Yellow => f.write_str("yellow"),
+            AnsiColor::Blue => f.write_str("blue"),
+            AnsiColor::Magenta => f.write_str("magenta"),
+            AnsiColor::Cyan => f.write_str("cyan"),
+            AnsiColor::White => f.write_str("white"),
+            AnsiColor::BrightBlack => f.write_str("bright-black"),
+            AnsiColor::BrightRed => f.write_str("bright-red"),
+            AnsiColor::BrightGreen => f.write_str("bright-green"),
+            AnsiColor::BrightYellow => f.write_str("bright-yellow"),
+            AnsiColor::BrightBlue => f.write_str("bright-blue"),
+            AnsiColor::BrightMagenta => f.write_str("bright-magenta"),
+            AnsiColor::BrightCyan => f.write_str("bright-cyan"),
+            AnsiColor::BrightWhite => f.write_str("bright-white"),
+        }
+    }
+}
+
+impl FromStr for Ansi {
+    type Err = AnsiParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
```

Comment:
is a lowercase match not easier to write?

Reply:
Yea, but I wanted to allow for an (allocation free) case-insensitive comparison so that during deserialization `BLACK` is also allowed (same as CSS).